docs: fixed faulty import in angular elements example

MrJean opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
The example A Popup Service containes a faulty import statement and use of function in `app.component.ts`, which results in a non working example.

Issue Number: N/A


## What is the new behavior?
The import statement including the function has been changed so this results in a working example.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

MrJean committed:
docs: fix `elements` example

MrJean commented:
@gkalpak Let me look into that. I'm quite new to contributing to the project so still trying to find my way round the code base.

gkalpak commented:
That's awesome, @MrJean! Feel free to ask here, if you have any questions or run into diffuculties or need more info/clarifications/guidance.

<sub>I know that figuring out the whole process as a first-time contributor can be overwhelming - we're here to help :smiley:</sub>

MrJean committed:
docs: add e2e tests for `elements` example

MrJean commented:
@gkalpak currently having trouble running a single e2e test with the command `yarn example-e2e --filter=elements`. It keeps running the general e2e tests and so I currently was not able to verify my e2e tests.

Also, for https://github.com/angular/angular/pull/24840/commits/0e73c74690ef0f4ec7233645529af55aa8f0a2a5 I changed the type of `popupEl` to `any` because my linter kept complaining `message` wasn't a known property of that element. Which is in fact the `@input()` of the component on the `HTMLElement`. I don't feel this is a proper solution and I wonder if this could be done better?

gkalpak commented on aio/content/examples/elements/e2e/src/app.e2e-spec.ts:
```diff
@@ -0,0 +1,28 @@
+      popupComponentButton.click();
+
+      const popupComponent =  element.all(by.tagName('popup-component')).get(0);
+      expect(popupComponent.innerHtml).toEqual('Popup: Message');
```
Is `innerHtml` a thing?

MrJean replied:
It seems it's not. I've replaced it with `getText()` https://www.protractortest.org/#/api?view=webdriver.WebElement.prototype.getText

gkalpak commented on aio/content/examples/elements/e2e/src/app.e2e-spec.ts:
```diff
@@ -0,0 +1,28 @@
+
+    it('should display popup component with content `Popup: Message`', function () {
+      const popupComponentButton = element.all(by.tagName('button')).get(0);
+      popupComponentButton.click();
```
Can you also verify that there is no `popup-component` before clicking the button (here and below)?

gkalpak reviewed (changes requested):
Thx for working on this, @MrJean. I've left some comments and I will look into the e2e test issue.

MrJean commented on aio/content/examples/elements/src/main.ts:
```diff
@@ -0,0 +1,12 @@
+// #docregion
```
Removed this

MrJean commented on aio/content/examples/elements/src/app/popup.service.ts:
```diff
@@ -40,7 +39,7 @@ export class PopupService {
   showAsElement(message: string) {
     // Create element
-    const popupEl = document.createElement('popup-element');
+    const popupEl: any = document.createElement('popup-element');
```
This is great! I did not know about `WithProperties`!

gkalpak reviewed (changes requested):
Almost there :smiley:
Can you, please, also take care of https://github.com/angular/angular/pull/24840#discussion_r202091680?

MrJean commented on aio/content/examples/elements/src/index.html:
```diff
@@ -0,0 +1,14 @@
+    <base href="/">
+    <title>Elements</title>
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <link rel="stylesheet" href="assets/sample.css">
```
@gkalpak I have removed this file

gkalpak replied:
It seems like you have removed the whole `index.html` file. I was only referring to this line:

```html
<link rel="stylesheet" href="assets/sample.css">
```

:grin:

MrJean commented on aio/tools/examples/shared/package.json:
```diff
@@ -22,6 +22,7 @@
     "@angular/common": "^6.0.0",
     "@angular/compiler": "^6.0.0",
     "@angular/core": "^6.0.0",
+    "@angular/elements": "^6.0.0",
```
@gkalpak That is right, forgot to commit it

MrJean commented:
@gkalpak Having some strange behaviour with the e2e tests. When checking wether the element is present, this seems to work by design (I hope so). But when I try to check the contents of the element it says the element with selector could not be found. I have added `xit` to the 2 tests so they don't break. Something tells me it has to do with the component/element processing the input and the browser not being ready yet?

gkalpak committed:
docs: clean up `elements` example (indentation, import order, etc)

gkalpak committed:
docs: refactor `elements` example

This makes the closing behavior more deterministic, which makes it
easier to be e2e-tested.

gkalpak committed:
fixup! docs: add e2e tests for `elements` example

gkalpak committed:
fixup! docs: add e2e tests for `elements` example

gkalpak commented:
I have pushed some changes to this PR:

- Did some minor refactoring in the `elements` example (mainly to resolve one of the testing issues - see below).
- Squashed your commits down to 2 (because you would need to do that anyway before merging).
- Added a couple of small fixes in `!fixup` commits. (These `!fixup` commits will be folded into yours while merging, so make sure you are OK with the contents :grin:)
- Addd a "work in progress" (WIP) commit with some test refactorings.

In case you're wondering there were two issues with the tests:

1. Since the popup would auto-hide after 2 secs (and since protractor tests are async), the popups would hide before being able to grab their text.
    _I fixed that by adding a "close" button and remove the auto-close-after-timeout behavior._
2. According to the spec, Custom Elements can only be used with ES2015 classes; i.e ES5 constructor functions won't work. (There are polyfills that allow using constructor functions.)
    By default, our examples are transpiled to ES5, which causes the popup element to break (the popup component works fine).
    _I still need to figure out the best way to fix this. We either need a special build for the `elements` example that transpiles to ES2015 or we need to include a polyfill. Off the top of my head, I lean towards the former for now and maybe later switch to polyfills and also explain the polyfill situation in the guide._

##
TODO:

- [ ] Fix issue 2 (see above).
- [ ] Add more tests (e.g. closing the popups, changing the message).

MrJean commented:
@gkalpak Great input from your side! 

1. I'll add more tests, but for know the `popup-element` will need to be placed in `xit` since they fail due to wrong build run (es2015 build needed or polyfill). Can you agree on this part?

2. I'd go for the es2015 build, but not sure where to start on that one.

MrJean committed:
fixup! docs: add e2e tests for `elements` example

MrJean commented:
@gkalpak While we're at it, I've added your explanation in https://github.com/angular/angular/pull/24840#issuecomment-405033821 to the docs.

gkalpak commented:
Thx for the changes, @MrJean. I will temporarily take this over, since some of the fixes require changes in our docs generation infrastructure, which is not trivial :smiley:

I'll hand it back over to you as soon as I sort these out. Hang tight 😄 
(And thanks for all the work so far ❤️)

gkalpak committed:
fixup! docs: add e2e tests for `elements` example

gkalpak committed:
build(docs-infra): add support for examples of type `elements`

Examples using `@angular/elements` need to transpile to es2015 for
Custom Elements to work (on browsers that natively support them).

Alternatively, a polyfill would need to be loaded. For now, changing the
transpilation target to es2015 is the simplest solution.

mary-poppins commented:
You can preview 959749f at https://pr24840-959749f.ngbuilds.io/.
You can preview 0f0d626 at https://pr24840-0f0d626.ngbuilds.io/.
You can preview 531ac8b at https://pr24840-531ac8b.ngbuilds.io/.
You can preview bf5ba31 at https://pr24840-bf5ba31.ngbuilds.io/.
You can preview 7484388 at https://pr24840-7484388.ngbuilds.io/.
You can preview df7f195 at https://pr24840-df7f195.ngbuilds.io/.
You can preview 2058cfb at https://pr24840-2058cfb.ngbuilds.io/.
You can preview dd7327a at https://pr24840-dd7327a.ngbuilds.io/.

mary-poppins commented:
You can preview 8f0d26a at https://pr24840-8f0d26a.ngbuilds.io/.

mary-poppins commented:
You can preview 086771d at https://pr24840-086771d.ngbuilds.io/.

gkalpak reviewed (approved):
I have some suggestions for changes in 086771d03 ("docs(elements): explanation about getting the typings right"), but since it isn't related to the purpose of this PR (fix broken syntax and add tests to ensure it doesn't break again), how about this:

1. Remove that commit from this PR.
2. Get this PR merged.
3. Open a new PR adding the typings info.

MrJean commented:
@gkalpak I've reverted the commit on the docs as asked.

MrJean commented:
@gkalpak New PR opened with doc content changes #24925

mary-poppins commented:
You can preview 845d8ad at https://pr24840-845d8ad.ngbuilds.io/.

gkalpak commented:
Instead of having both the initial commit and the revert commit (which will be confusing for people looking at the commit history in the future), it would be better to just remove both commits altogether.

You can do this with `git rebase --interactive ...` (more info on [rewriting history][1]).
Or if these are the last two commits, you can `git reset --hard @~2`. 
(In both case you need to `git push --force-with-lease` once you are done.)

[1]: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

MrJean commented:
@gkalpak I've misunderstood what had to be done. This should be OK now.

mary-poppins commented:
You can preview cb55d40 at https://pr24840-cb55d40.ngbuilds.io/.

mary-poppins commented:
You can preview cde68d6 at https://pr24840-cde68d6.ngbuilds.io/.

gkalpak commented:
Note to caretaker:
The Travis failure is due to flakey Saucelabs and not related to this PR.

MrJean commented:
Any idea why the CLA got unsigned?

gkalpak commented:
Thx, @matsko, for taking care of the CLA.

##
@MrJean, [this commit][1] pretty much explains the issue. If a PR includes commits by people other than the PR submitter (e.g. your PR has my commits), googlebot can't verify that I'm OK with you submitting my commits in your PR (it doesn't know I added them), so we have to manually verify that it's OK.
(But the god news is it was able to verify both our CLAs :wink:)

BTW, great work on this one :100: It should be merged and live soon :tada:

[1]: https://github.com/angular/angular/pull/24840#issuecomment-405043690

mathpaquette commented:
@MrJean @jbogarthyde will this example has a memory leak without removeEventListener to the 'closed' event?

gkalpak commented:
@mathpaquette, your question is a little vague, but if I understand it correctly, there won't be a memory leak, becasue the element is removed from the DOM and (if there are no other references to it) it will be garbage collected (along with its event listeners).
